src/Microsoft.IdentityModel.Protocols/Configuration/ConfigurationManager.cs

```diff
@@ -144,83 +149,174 @@ public async Task<T> GetConfigurationAsync()
+
+                    // In this case configuration was never obtained.
+                    if (_currentConfiguration == null)
+                    {
+                        if (_bootstrapRefreshInterval < RefreshInterval)
+                        {
+                            // Adopt exponential backoff for bootstrap refresh interval with a decorrelated jitter if it is not longer than the refresh interval.
+                            TimeSpan _bootstrapRefreshIntervalWithJitter = TimeSpan.FromSeconds(new Random().Next((int)_bootstrapRefreshInterval.TotalSeconds));
+                            _bootstrapRefreshInterval += _bootstrapRefreshInterval;
+                            _syncAfter = DateTimeUtil.Add(DateTime.UtcNow, _bootstrapRefreshIntervalWithJitter);
                         }
                         else
                         {
-                            _syncAfter = DateTimeUtil.Add(DateTime.UtcNow, AutomaticRefreshInterval < RefreshInterval ? AutomaticRefreshInterval : RefreshInterval);
-
-                            LogHelper.LogExceptionMessage(
-                                new InvalidOperationException(
-                                    LogHelper.FormatInvariant(LogMessages.IDX20806, LogHelper.MarkAsNonPII(MetadataAddress ?? "null"), LogHelper.MarkAsNonPII(ex)), ex));
+                            _syncAfter = DateTimeUtil.Add(
+                                DateTime.UtcNow,
+                                AutomaticRefreshInterval < RefreshInterval ? AutomaticRefreshInterval : RefreshInterval);
                         }
+
+                        throw LogHelper.LogExceptionMessage(
+                            new InvalidOperationException(
+                                LogHelper.FormatInvariant(
+                                    LogMessages.IDX20803,
+                                    LogHelper.MarkAsNonPII(MetadataAddress ?? "null"),
+                                    LogHelper.MarkAsNonPII(_syncAfter),
+                                    LogHelper.MarkAsNonPII(ex)),
+                                ex));
                     }
+                    else
+                    {
+                        _syncAfter = DateTimeUtil.Add(
+                            DateTime.UtcNow,
+                            AutomaticRefreshInterval < RefreshInterval ? AutomaticRefreshInterval : RefreshInterval);
+
+                        LogHelper.LogExceptionMessage(
+                            new InvalidOperationException(
+                                LogHelper.FormatInvariant(
+                                    LogMessages.IDX20806,
+                                    LogHelper.MarkAsNonPII(MetadataAddress ?? "null"),
+                                    LogHelper.MarkAsNonPII(ex)),
+                                ex));
+                    }
+                }
+                finally
+                {
+                    _configurationNullLock.Release();
```

Comment:
this could be in a much narrower try finally block so on the typical path after the first fetch the try finally is never entered.

Reply:
@keegan-caruso 
True, but in the abundance of caution, if the finally is not entered, the code can never recover.

Reply:
Ah, I'm sorry, I read the diff wrong. I follow now.